File: src/main/utils/sigv4.ts
import { createHash, createHmac } from 'crypto';

/**
 * Minimal AWS Signature Version 4 signer (no SDK dependency).
 *
 * Covers the case Echo needs: a single signed POST/GET to a regional AWS
 * endpoint with static credentials. The path is treated as already canonical
 * (used verbatim for both the canonical request and the wire request), which
 * matches the AWS sig-v4 test-suite `get-vanilla` vector and how Bedrock model
 * paths (e.g. `/model/anthropic.claude-...:0/invoke`) are sent.
 *
 * Verified against the official AWS test vector in tests/sigv4.test.ts.
 */
export interface SigV4Request {
  method: string;
  host: string;
  /** Absolute path, used verbatim as the canonical URI and the request path. */
  path: string;
  region: string;
  service: string;
  /** Request payload as a string ('' for empty bodies). */
  body: string;
  accessKeyId: string;
  secretAccessKey: string;
  sessionToken?: string;
  /** Extra headers to include in the signature AND send (e.g. content-type). */
  extraHeaders?: Record<string, string>;
  /** Override the timestamp (YYYYMMDDTHHMMSSZ). For testing/determinism. */
  amzDate?: string;
}

function sha256Hex(data: string): string {
  return createHash('sha256').update(data, 'utf8').digest('hex');
}

function hmac(key: Buffer | string, data: string): Buffer {
  return createHmac('sha256', key).update(data, 'utf8').digest();
}

function formatAmzDate(d: Date): string {
  // '2015-08-30T12:36:00.000Z' -> '20150830T123600Z'
  return d.toISOString().replace(/[:-]|\.\d{3}/g, '');
}

/**
 * Returns the headers to send with the request, including `Authorization`,
 * `X-Amz-Date`, and any `extraHeaders`. The caller sends these on the wire and
 * lets the HTTP client add the `Host` header (which is part of the signature).
 */
export function signRequest(req: SigV4Request): Record<string, string> {
  const amzDate = req.amzDate ?? formatAmzDate(new Date());
  const dateStamp = amzDate.slice(0, 8);

  // Headers that participate in the signature (names lowercased, sorted).
  const signHeaders: Record<string, string> = {
    host: req.host,
    'x-amz-date': amzDate,
  };
  for (const [k, v] of Object.entries(req.extraHeaders ?? {})) {
    signHeaders[k.toLowerCase()] = v;
  }
  if (req.sessionToken) signHeaders['x-amz-security-token'] = req.sessionToken;

  const sortedNames = Object.keys(signHeaders).sort();
  const canonicalHeaders = sortedNames.map(n => `${n}:${signHeaders[n].trim()}\n`).join('');
  const signedHeaders = sortedNames.join(';');
  const payloadHash = sha256Hex(req.body);

  const canonicalRequest = [
    req.method,
    req.path,
    '', // canonical query string (none)
    canonicalHeaders,
    signedHeaders,
    payloadHash,
  ].join('\n');

  const scope = `${dateStamp}/${req.region}/${req.service}/aws4_request`;
  const stringToSign = [
    'AWS4-HMAC-SHA256',
    amzDate,
    scope,
    sha256Hex(canonicalRequest),
  ].join('\n');

  const kDate = hmac('AWS4' + req.secretAccessKey, dateStamp);
  const kRegion = hmac(kDate, req.region);
  const kService = hmac(kRegion, req.service);
  const kSigning = hmac(kService, 'aws4_request');
  const signature = createHmac('sha256', kSigning).update(stringToSign, 'utf8').digest('hex');

  const out: Record<string, string> = {
    'X-Amz-Date': amzDate,
    'Authorization': `AWS4-HMAC-SHA256 Credential=${req.accessKeyId}/${scope}, SignedHeaders=${signedHeaders}, Signature=${signature}`,
    ...(req.extraHeaders ?? {}),
  };
  if (req.sessionToken) out['X-Amz-Security-Token'] = req.sessionToken;
  return out;
}

File: src-tauri/src/utils/sigv4.rs
// Minimal AWS Signature Version 4 signer (no AWS SDK dependency).
//
// Mirror of src/main/utils/sigv4.ts. Covers a single signed POST/GET to a
// regional AWS endpoint with static credentials. The path is used verbatim as
// the canonical URI (matching the AWS sig-v4 `get-vanilla` vector and how
// Bedrock model paths are sent). Verified by the test at the bottom of this
// file (`cargo test sigv4`).

use hmac::{Hmac, Mac};
use sha2::{Digest, Sha256};

type HmacSha256 = Hmac<Sha256>;

pub struct SigV4Request<'a> {
    pub method: &'a str,
    pub host: &'a str,
    /// Absolute path, used verbatim as the canonical URI and request path.
    pub path: &'a str,
    pub region: &'a str,
    pub service: &'a str,
    /// Request payload ("" for empty bodies).
    pub body: &'a str,
    pub access_key_id: &'a str,
    pub secret_access_key: &'a str,
    /// Content-Type to sign AND send (e.g. "application/json").
    pub content_type: Option<&'a str>,
    /// Override the timestamp (YYYYMMDDTHHMMSSZ). For testing/determinism.
    pub amz_date: Option<&'a str>,
}

fn sha256_hex(data: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(data);
    hex::encode(hasher.finalize())
}

fn hmac_sha256(key: &[u8], data: &[u8]) -> Vec<u8> {
    let mut mac = HmacSha256::new_from_slice(key).expect("HMAC accepts any key length");
    mac.update(data);
    mac.finalize().into_bytes().to_vec()
}

/// Returns the headers to send (Authorization, X-Amz-Date, and Content-Type if
/// provided). The caller sends these and lets the HTTP client add the Host
/// header (which is part of the signature).
pub fn sign_request(req: &SigV4Request) -> Vec<(String, String)> {
    let amz_date = match req.amz_date {
        Some(d) => d.to_string(),
        None => chrono::Utc::now().format("%Y%m%dT%H%M%SZ").to_string(),
    };
    let date_stamp = &amz_date[..8];

    // Headers that participate in the signature (lowercased names, sorted).
    let mut headers: Vec<(String, String)> = vec![
        ("host".to_string(), req.host.to_string()),
        ("x-amz-date".to_string(), amz_date.clone()),
    ];
    if let Some(ct) = req.content_type {
        headers.push(("content-type".to_string(), ct.to_string()));
    }
    headers.sort_by(|a, b| a.0.cmp(&b.0));

    let canonical_headers: String = headers
        .iter()
        .map(|(n, v)| format!("{}:{}\n", n, v.trim()))
        .collect();
    let signed_headers = headers
        .iter()
        .map(|(n, _)| n.clone())
        .collect::<Vec<_>>()
        .join(";");
    let payload_hash = sha256_hex(req.body.as_bytes());

    let canonical_request = format!(
        "{}\n{}\n{}\n{}\n{}\n{}",
        req.method, req.path, "", canonical_headers, signed_headers, payload_hash
    );

    let scope = format!("{}/{}/{}/aws4_request", date_stamp, req.region, req.service);
    let string_to_sign = format!(
        "AWS4-HMAC-SHA256\n{}\n{}\n{}",
        amz_date,
        scope,
        sha256_hex(canonical_request.as_bytes())
    );

    let k_date = hmac_sha256(format!("AWS4{}", req.secret_access_key).as_bytes(), date_stamp.as_bytes());
    let k_region = hmac_sha256(&k_date, req.region.as_bytes());
    let k_service = hmac_sha256(&k_region, req.service.as_bytes());
    let k_signing = hmac_sha256(&k_service, b"aws4_request");
    let signature = hex::encode(hmac_sha256(&k_signing, string_to_sign.as_bytes()));

    let authorization = format!(
        "AWS4-HMAC-SHA256 Credential={}/{}, SignedHeaders={}, Signature={}",
        req.access_key_id, scope, signed_headers, signature
    );

    let mut out = vec![
        ("X-Amz-Date".to_string(), amz_date),
        ("Authorization".to_string(), authorization),
    ];
    if let Some(ct) = req.content_type {
        out.push(("Content-Type".to_string(), ct.to_string()));
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn get_vanilla_vector() {
        // Official AWS sig-v4-test-suite `get-vanilla` vector.
        let headers = sign_request(&SigV4Request {
            method: "GET",
            host: "example.amazonaws.com",
            path: "/",
            region: "us-east-1",
            service: "service",
            body: "",
            access_key_id: "AKIDEXAMPLE",
            secret_access_key: "wJalrXUtnFEMI/K7MDENG+bPxRfiCYEXAMPLEKEY",
            content_type: None,
            amz_date: Some("20150830T123600Z"),
        });
        let auth = headers.iter().find(|(k, _)| k == "Authorization").map(|(_, v)| v.clone()).unwrap();
        assert_eq!(
            auth,
            "AWS4-HMAC-SHA256 Credential=AKIDEXAMPLE/20150830/us-east-1/service/aws4_request, \
             SignedHeaders=host;x-amz-date, \
             Signature=5fa00fa31553b73ebf1942676e86291e8372ff2a2260956d9b8aae1d763fbf31"
        );
    }
}
